Sentiment Analysis for Online Reviews

In [21]:
import pandas as pd
import nltk as nltk
nltk.download('stopwords')
nltk.download('punkt')
from nltk.corpus import stopwords 
from nltk.tokenize import word_tokenize 

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\evatr\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\evatr\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.


a) Downloading, reading and analyzing datasets

In [105]:
# load data in the right> format according to readme files
yelp=pd.read_csv("sentiment_labelled_sentences\yelp_labelled.txt",delimiter="\t", names=["Sentence", "Label"])
imdb=pd.read_csv("sentiment_labelled_sentences\imdb_labelled.txt",delimiter="\t", names=["Sentence", "Label"])
amazon=pd.read_csv("sentiment_labelled_sentences\labelled_amazon.txt",delimiter="\t", names=["Sentence", "Label"])

In [7]:
# check if data is balance in all three dataframes

# yelp
onesYelp = len(yelp[yelp['Label'] == 1])
zerosYelp = len(yelp[yelp['Label'] == 0])
print('Number of 1s in Yelp:', onesYelp)
print('Number of 0s in Yelp:', zerosYelp)

#imdb
onesImdb = len(imdb[imdb['Label'] == 1])
zerosImdb = len(imdb[imdb['Label'] == 0])
print('Number of 1s in Imdb:', onesImdb)
print('Number of 0s in Imdb:', zerosImdb)

#amazon
onesAmazon = len(amazon[amazon['Label'] == 1])
zerosAmazon = len(amazon[amazon['Label'] == 0])
print('Number of 1s in Amazon:', onesAmazon)
print('Number of 0s in Amazon:', zerosAmazon)

Number of 1s in Yelp: 500
Number of 0s in Yelp: 500
Number of 1s in Imdb: 386
Number of 0s in Imdb: 362
Number of 1s in Amazon: 500
Number of 0s in Amazon: 500


The data in the Yelp and Amazon files is balanced because there are the same number of 1s and 0s as labels. 
The data in the Imdb file can be considered almost balancen because the number of 1s and 0s is almost the same (386 and 362, respectively). The ratio of 1s to 0s is 386/362 = 1.067.

b) Pre-processing datasets

In [106]:
# convert all letters to lower case
yelp = yelp.applymap(lambda s:s.lower() if type(s) == str else s)
imdb = imdb.applymap(lambda s:s.lower() if type(s) == str else s)
amazon = amazon.applymap(lambda s:s.lower() if type(s) == str else s)

# remove stop words

stopWords = set(stopwords.words('english')) # find stop words in English language
strippedSentence = [] # list to store the new sentence stripped of stop words

# iterate through every sentence and replace it by itself without stop words - CHECK WHY OUTPUT APPEARS WITH COMMAS
for i in yelp['Sentence'].index:
    wordTokens = word_tokenize(yelp.at[i, 'Sentence'])
    for word in wordTokens: 
        if word not in stopWords: 
            strippedSentence.append(word)
    (yelp.at[i, 'Sentence']) = strippedSentence
    strippedSentence = []

print(yelp['Sentence']) # check it worked correctly

0                            [wow, ..., loved, place, .]
1                                       [crust, good, .]
2                             [tasty, texture, nasty, .]
3      [stopped, late, may, bank, holiday, rick, stev...
4                    [selection, menu, great, prices, .]
5                   [getting, angry, want, damn, pho, .]
6                    [honeslty, n't, taste, fresh, ., )]
7      [potatoes, like, rubber, could, tell, made, ah...
8                                      [fries, great, .]
9                                      [great, touch, .]
10                                  [service, prompt, .]
11                                  [would, go, back, .]
12     [cashier, care, ever, say, still, ended, wayyy...
13     [tried, cape, cod, ravoli, ,, chicken, ,, cran...
14             [disgusted, pretty, sure, human, hair, .]
15                   [shocked, signs, indicate, cash, .]
16                              [highly, recommended, .]
17                  [waitress, 

For this part, we decided to convert all sentences to lower case, so that the same word with some upper case letters and without them would not be detected as different words since we are using the string type which takes into account their differences. We also stripped the sentences of stop words because they do not add any meaning as the same stop words appear in many different sentences.

c) Split training and testing data

In [78]:
# not finished / working yet

training_yelp = (yelp.query('Label == 1' )).head(400)
training_yelp.append((yelp.query('Label == 0' )).head(400))
testing_yelp = (yelp.query('Label == 1' )).tail(100)
testing_yelp.append((yelp.query('Label == 0' )).tail(100))

Sentence  Label
715  only pros : large seating area/ nice bar area/...      1
716                they have a really nice atmosphere.      1
718                      after one bite, i was hooked.      1
720                      cute, quaint, simple, honest.      1
721  the chicken was deliciously seasoned and had t...      1
722  the food was great as always, compliments to t...      1
723  special thanks to dylan t. for the recommendat...      1
724                         awesome selection of beer.      1
725                    great food and awesome service!      1
726  one nice thing was that they added gratuity on...      1
728           the han nan chicken was also very tasty.      1
729         as for the service, i thought it was good.      1
732    nicest chinese restaurant i've been in a while.      1
733        overall, i like there food and the service.      1
734  they also now serve indian naan bread with hum...      1
737  try them in the airport to experience some tas...      1
738  i love the decor with the chinese calligraphy ...      1
739         never had anything to complain about here.      1
740  the restaurant is very clean and has a family ...      1
745  there was a warm feeling with the service and ...      1
746  an extensive menu provides lots of options for...      1
747  i always order from the vegetarian menu during...      1
749  wonderful lil tapas and the ambience made me f...      1
750  i got to enjoy the seafood salad, with a fabul...      1
751  the wontons were thin, not thick and chewy, al...      1
752  level 5 spicy was perfect, where spice didn't ...      1
753  we were sat right on time and our server from ...      1
755  when i'm on this side of town, this will defin...      1
757                      this is a great place to eat!      1
758                  the jalapeno bacon is soooo good.      1
..                                                 ...    ...
970  i immediately said i wanted to talk to the man...      0
971                    the ambiance isn't much better.      0
972  unfortunately, it only set us up for disapppoi...      0
973                              the food wasn't good.      0
974  your servers suck, wait, correction, our serve...      0
975      what happened next was pretty....off putting.      0
976  too bad cause i know it's family owned, i real...      0
977               overpriced for what you are getting.      0
978               i vomited in the bathroom mid lunch.      0
979  i kept looking at the time and it had soon bec...      0
980  i have been to very few places to eat that und...      0
981  we started with the tuna sashimi which was bro...      0
982                            food was below average.      0
983  it sure does beat the nachos at the movies but...      0
984       all in all, ha long bay was a bit of a flop.      0
985  the problem i have is that they charge $11.99 ...      0
986  shrimp- when i unwrapped it (i live only 1/2 a...      0
987     it lacked flavor, seemed undercooked, and dry.      0
988  it really is impressive that the place hasn't ...      0
989  i would avoid this place if you are staying in...      0
990  the refried beans that came with my meal were ...      0
991         spend your money and time some place else.      0
992  a lady at the table next to us found a live gr...      0
993            the presentation of the food was awful.      0
994           i can't tell you how disappointed i was.      0
995  i think food should have flavor and texture an...      0
996                           appetite instantly gone.      0
997  overall i was not impressed and would not go b...      0
998  the whole experience was underwhelming, and i ...      0
999  then, as if i hadn't wasted enough of my life ...      0

[200 rows x 2 columns]

In [79]:
training_yelp

Sentence  Label
0                             wow... loved this place.      1
3    stopped by during the late may bank holiday of...      1
4    the selection on the menu was great and so wer...      1
8                            the fries were great too.      1
9                                       a great touch.      1
10                            service was very prompt.      1
13   i tried the cape cod ravoli, chicken,with cran...      1
16                                 highly recommended.      1
21                                  the food, amazing.      1
22                               service is also cute.      1
23   i could care less... the interior is just beau...      1
24                                  so they performed.      1
25   that's right....the red velvet cake.....ohhh t...      1
27   this hole in the wall has great mexican street...      1
30   also there are combos like a burger, fries, an...      1
32   i found this place by accident and i could not...      1
34                   overall, i like this place a lot.      1
35   the only redeeming quality of the restaurant w...      1
36                     ample portions and good prices.      1
38               my first visit to hiro was a delight!      1
40                        the shrimp tender and moist.      1
43   on a positive note, our server was very attent...      1
45   the only thing i did like was the prime rib an...      1
47         the burger is good beef, cooked just right.      1
48   if you want a sandwich just go to any firehous...      1
49   my side greek salad with the greek dressing wa...      1
50   we ordered the duck rare and it was pink and t...      1
51   he came running after us when he realized my h...      1
52                         their chow mein is so good!      1
54                               the portion was huge!      1
..                                                 ...    ...
665  the sweet potato tots were good but the onion ...      1
666                      the staff was very attentive.      1
667  and the chef was generous with his time (even ...      1
668  the owner used to work at nobu, so this place ...      1
671                   i promise they won't disappoint.      1
673                  what a great double cheeseburger!      1
674                          awesome service and food.      1
675                   a fantastic neighborhood gem !!!      1
676                           i can't wait to go back.      1
678      it's a great place and i highly recommend it.      1
680  i gave it 5 stars then, and i'm giving it 5 st...      1
682                  dessert: panna cotta was amazing.      1
683                 very good food, great atmosphere.1      1
684                                   damn good steak.      1
686  prices are very reasonable, flavors are spot o...      1
687  the decor is nice, and the piano music soundtr...      1
688  the steak was amazing...rge fillet relleno was...      1
689                         good food , good service .      1
690                         it was absolutely amazing.      1
692                           will definitely be back!      1
693  the sergeant pepper beef sandwich with auju sa...      1
694  hawaiian breeze, mango magic, and pineapple de...      1
702  have been going since 2007 and every meal has ...      1
703  our server was very nice and attentive as were...      1
704  the cashier was friendly and even brought the ...      1
706                        the atmosphere here is fun.      1
708  service is quick and even "to go" orders are j...      1
710  that said, our mouths and bellies were still q...      1
712                                      2 thumbs up!!      1
714  i loved the grilled pizza, reminded me of legi...      1

[400 rows x 2 columns]